runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PTransformTranslation.java

```diff
@@ -179,7 +179,10 @@ private PTransformTranslation() {}
     if (transform instanceof RawPTransform) {
       // The raw transform was parsed in the context of other components; this puts it in the
       // context of our current serialization
-      transformBuilder.setSpec(((RawPTransform<?, ?>) transform).migrate(components));
+      FunctionSpec spec = ((RawPTransform<?, ?>) transform).migrate(components);
+      if (spec != null) {
```

Comment:
Maybe add a brief comment why null specs are fine?

Reply:
TBH I'm on the fence about it. Just that GBK and Flatten could have dummy specs and we might as well save the typing and bandwidth...

Reply:
Sorry, I was conflating spec and payload. It is OK to have a null spec or a URN with a null payload.